File: 3.x.x/crates-utopia/dns/src/message/question.rs
use crate::error::{Error, Result};
use crate::message::domain::Domain;
use crate::message::record::Record;
use crate::wire::{normalize_name, push_u16, read_u16};

/// DNS question. PHP `Utopia\DNS\Message\Question`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Question {
    pub name: String,
    pub type_code: u16,
    pub class: u16,
}

impl Question {
    /// PHP `Question::__construct`. Name is trimmed and lowercased.
    #[must_use]
    pub fn new(name: impl AsRef<str>, type_code: u16) -> Self {
        Self::with_class(name, type_code, Record::CLASS_IN)
    }

    #[must_use]
    pub fn with_class(name: impl AsRef<str>, type_code: u16, class: u16) -> Self {
        Self {
            name: normalize_name(name.as_ref()),
            type_code,
            class,
        }
    }

    /// PHP `Question::decode`.
    pub fn decode(data: &[u8], offset: &mut usize) -> Result<Self> {
        let name = Domain::decode(data, offset)?;
        let remaining = data.len().saturating_sub(*offset);
        if remaining < 4 {
            return Err(Error::decoding("Question section truncated"));
        }
        let type_code = read_u16(data, *offset)
            .map_err(|_| Error::decoding("Failed to unpack question type"))?;
        *offset += 2;
        let class = read_u16(data, *offset)
            .map_err(|_| Error::decoding("Failed to unpack question class"))?;
        *offset += 2;
        Ok(Self::with_class(name, type_code, class))
    }

    /// PHP `Question::encode`.
    pub fn encode(&self) -> Result<Vec<u8>> {
        let mut data = Domain::encode(&self.name)?;
        push_u16(&mut data, self.type_code);
        push_u16(&mut data, self.class);
        Ok(data)
    }
}

File: packages/dns/src/Message/Question.php
<?php

namespace Utopia\DNS\Message;

use Utopia\DNS\Exception\Message\DecodingException;

final readonly class Question
{
    public string $name;

    public function __construct(
        string $name,
        public int $type,
        public int $class = Record::CLASS_IN,
    ) {
        $this->name = trim(strtolower($name));
    }

    /**
     * @param-out int $offset
     */
    public static function decode(string $data, int &$offset = 0): self
    {
        $name = Domain::decode($data, $offset);

        $remaining = \strlen($data) - $offset;
        if ($remaining < 4) {
            throw new DecodingException('Question section truncated');
        }

        $typeData = unpack('ntype', substr($data, $offset, 2));
        if (!\is_array($typeData) || !\array_key_exists('type', $typeData) || !\is_int($typeData['type'])) {
            throw new DecodingException('Failed to unpack question type');
        }
        $type = $typeData['type'];
        $offset += 2;

        $classData = unpack('nclass', substr($data, $offset, 2));
        if (!\is_array($classData) || !\array_key_exists('class', $classData) || !\is_int($classData['class'])) {
            throw new DecodingException('Failed to unpack question class');
        }
        $class = $classData['class'];
        $offset += 2;

        return new self($name, $type, $class);
    }

    public function encode(): string
    {
        $encodedName = Domain::encode($this->name);

        return $encodedName . pack('nn', $this->type, $this->class);
    }
}
